### PR TITLE
Controller should check authorization for /debug/pprof

### PR DESCRIPTION
**Pull request by smarterclayton**
Make standalone controllers be debuggable securely by exposing the pprof handlers and adding remote authz and authn filters with small caches.

Also fixes SAR to support Non resource URLs like upstream.

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
@liggitt 

Still need to make the change to have a short-circuit authorizer for /healthz (which always needs to be unauthenticated).

File: pkg/authorization/api/validation/validation.go [hunk 1]
```diff
@@ -35,7 +35,14 @@ func ValidateSubjectAccessReview(review *authorizationapi.SubjectAccessReview) f
 	if len(review.Action.Verb) == 0 {
 		allErrs = append(allErrs, field.Required(field.NewPath("verb"), ""))
 	}
-	if len(review.Action.Resource) == 0 {
+	if review.Action.IsNonResourceURL {
+		if len(review.Action.Path) == 0 {
+			allErrs = append(allErrs, field.Required(field.NewPath("path"), ""))
+		}
+		if len(review.Action.Resource) != 0 {
```
Comment (liggitt) [inline pkg/authorization/api/validation/validation.go]: disallow all the resource-related fields (namespace, group, resource, resourcename) for non-resource urls

File: pkg/authorization/api/validation/validation.go [hunk 2]
```diff
@@ -48,7 +55,14 @@ func ValidateResourceAccessReview(review *authorizationapi.ResourceAccessReview)
 	if len(review.Action.Verb) == 0 {
 		allErrs = append(allErrs, field.Required(field.NewPath("verb"), ""))
 	}
-	if len(review.Action.Resource) == 0 {
+	if review.Action.IsNonResourceURL {
```
Comment (liggitt) [inline pkg/authorization/api/validation/validation.go]: heh, that makes ResourceAccessReview slightly nonsensical. oh well

File: pkg/authorization/api/validation/validation.go [hunk 3]
```diff
@@ -48,7 +55,14 @@ func ValidateResourceAccessReview(review *authorizationapi.ResourceAccessReview)
...
+		if len(review.Action.Path) == 0 {
+			allErrs = append(allErrs, field.Required(field.NewPath("path"), ""))
+		}
+		if len(review.Action.Resource) != 0 {
```
Comment (liggitt) [inline pkg/authorization/api/validation/validation.go]: prevent all resource-related attributes

File: pkg/authorization/api/validation/validation.go [hunk 4]
```diff
@@ -61,7 +75,14 @@ func ValidateLocalSubjectAccessReview(review *authorizationapi.LocalSubjectAcces
...
```
Comment (liggitt) [inline pkg/authorization/api/validation/validation.go]: prevent all resource-related attributes

File: pkg/authorization/api/validation/validation.go [hunk 5]
```diff
@@ -74,7 +95,14 @@ func ValidateLocalResourceAccessReview(review *authorizationapi.LocalResourceAcc
...
```
Comment (liggitt) [inline pkg/authorization/api/validation/validation.go]: prevent all resource-related attributes

File: pkg/cmd/server/handlers/authentication.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/auth/authenticator"
+)
+
+// AuthenticationHandlerFilter creates a filter object that will enforce authentication directly
+func AuthenticationHandlerFilter(handler http.Handler, authenticator authenticator.Request, contextMapper kapi.RequestContextMapper) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		user, ok, err := authenticator.AuthenticateRequest(req)
+		if err != nil || !ok {
+			http.Error(w, "Unauthorized", http.StatusUnauthorized)
```
Comment (liggitt) [inline pkg/cmd/server/handlers/authentication.go]: huh, hope everyone deals with non-API responses well

File: pkg/cmd/server/origin/auth.go [hunk 7]
```diff
@@ -762,27 +762,3 @@ func (redirectSuccessHandler) AuthenticationSucceeded(user kuser.Info, then stri
 	http.Redirect(w, req, then, http.StatusFound)
 	return true, nil
 }
-
-// authenticationHandlerFilter creates a filter object that will enforce authentication directly
-func authenticationHandlerFilter(handler http.Handler, authenticator authenticator.Request, contextMapper kapi.RequestContextMapper) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		user, ok, err := authenticator.AuthenticateRequest(req)
-		if err != nil || !ok {
-			http.Error(w, "Unauthorized", http.StatusUnauthorized)
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: surprised we were returning this from the API... I thought only the node did this

**Comment by liggitt** [comment]
assuming the big blocks were straight moves, seems mostly sane. still not quite sure about controller manager as a serviceaccount

**Comment by smarterclayton** [comment]
I think skip controllerm as service account for now.  Doesn't help in short
term and long term unclear.

On Feb 10, 2017, at 1:35 AM, Jordan Liggitt <notifications@github.com>
wrote:

assuming the big blocks were straight moves, seems mostly sane. still not
quite sure about controller manager as a serviceaccount

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/12907#issuecomment-278869094>,
or mute
the thread

**Comment by smarterclayton** [comment]
Yeah, surprised me as well.  Should be negotiated now, will need to check
with @sttts on upstream code

On Feb 10, 2017, at 1:32 AM, Jordan Liggitt <notifications@github.com>
wrote:

*@liggitt* commented on this pull request.
------------------------------

In pkg/cmd/server/origin/auth.go
<https://github.com/openshift/origin/pull/12907#pullrequestreview-21186297>:

> @@ -762,27 +762,3 @@ func (redirectSuccessHandler) AuthenticationSucceeded(user kuser.Info, then stri
 	http.Redirect(w, req, then, http.StatusFound)
 	return true, nil
 }
-
-// authenticationHandlerFilter creates a filter object that will
enforce authentication directly
-func authenticationHandlerFilter(handler http.Handler, authenticator
authenticator.Request, contextMapper kapi.RequestContextMapper)
http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		user, ok, err := authenticator.AuthenticateRequest(req)
-		if err != nil || !ok {
-			http.Error(w, "Unauthorized", http.StatusUnauthorized)

surprised we were returning this from the API... I thought only the node
did this

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/12907#pullrequestreview-21186297>,
or mute the thread

File: pkg/cmd/server/origin/auth.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/auth.go]: same

**Comment by smarterclayton** [comment]
All big blocks were moves.

**Comment by smarterclayton** [comment]
Ok, added a quick test and the /healthz and /healthz/ready bypass and tested locally.

**Comment by stevekuznetsov** [comment]
[test]

File: test/cmd/authentication.sh [hunk 8]
```diff
@@ -77,7 +78,20 @@ os::cmd::expect_success_and_text "oc policy can-i create pods --token='${accesst
...
 os::cmd::expect_success_and_not_text "oc policy can-i --list --token='${accesstoken}' -n '${project}'" 'get.*pods'
 
+os::test::junit::declare_suite_end
 
+os::test::junit::declare_suite_start "cmd/authentication/debugging"
+os::cmd::expect_success_and_text 'oc login -u user3 -p pw' 'Login successful'
+os::cmd::expect_success 'oc login -u system:admin'
+os::cmd::expect_failure_and_text 'oc get --raw /debug/pprof/ --as=user3' 'Forbidden'
+os::cmd::expect_failure_and_text 'oc get --raw /metrics --as=user3' 'Forbidden'
+os::cmd::expect_success 'oadm policy add-cluster-role-to-user cluster-debugger user3'
```
Comment (liggitt) [inline test/cmd/authentication.sh]: needs wait_for to avoid flakes in the gets below

File: test/cmd/authentication.sh [hunk 9]
```diff
@@ -77,7 +78,20 @@ os::cmd::expect_success_and_text "oc policy can-i create pods --token='${accesst
 os::cmd::expect_success_and_text "oc policy can-i --list --token='${accesstoken}' -n '${project}' --scopes='role:admin:*'" 'get.*pods'
 os::cmd::expect_success_and_not_text "oc policy can-i --list --token='${accesstoken}' -n '${project}'" 'get.*pods'
 
+os::test::junit::declare_suite_end
 
+os::test::junit::declare_suite_start "cmd/authentication/debugging"
+os::cmd::expect_success_and_text 'oc login -u user3 -p pw' 'Login successful'
+os::cmd::expect_success 'oc login -u system:admin'
+os::cmd::expect_failure_and_text 'oc get --raw /debug/pprof/ --as=user3' 'Forbidden'
+os::cmd::expect_failure_and_text 'oc get --raw /metrics --as=user3' 'Forbidden'
```
Comment (liggitt) [inline test/cmd/authentication.sh]: test /healthz access is allowed even without the role?

File: pkg/cmd/server/origin/master.go [hunk 10]
```diff
@@ -328,14 +331,38 @@ func (c *MasterConfig) buildHandlerChain(assetConfig *AssetConfig) (func(http.Ha
...
+	if err != nil {
+		return err
+	}
+	// we use direct bypass to allow readiness and health to work regardless of the master health
+	authz = serverhandlers.NewBypassAuthorizer(authz, "/healthz", "/healthz/ready")
+	contextMapper := c.getRequestContextMapper()
+	handler := serverhandlers.AuthorizationFilter(apiContainer.ServeMux, authz, c.AuthorizationAttributeBuilder, contextMapper)
+	handler = serverhandlers.AuthenticationHandlerFilter(handler, authn, contextMapper)
+	handler = kgenericfilters.WithPanicRecovery(handler, contextMapper)
+	handler = kapiserverfilters.WithRequestInfo(handler, genericapiserver.NewRequestInfoResolver(&genericapiserver.Config{}), contextMapper)
```
Comment (liggitt) [inline pkg/cmd/server/origin/master.go]: is this request info resolver going to try to build API attributes if the path starts with /apis, etc?

**Comment by liggitt** [comment]
a few nits, LGTM

**Comment by stevekuznetsov** [comment]
Not a flake:
```
[WARNING] While the jUnit report found no failed tests, the `go test` process failed.
[WARNING] This usually means that the unit test suite failed to compile.
[WARNING] `go test` had the following output to stderr:
# github.com/openshift/origin/pkg/cmd/server/origin
pkg/cmd/server/origin/handlers_test.go:289: config.impersonationFilter undefined (type MasterConfig has no field or method impersonationFilter)
```

File: pkg/cmd/server/origin/master.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: yes it will

**Comment by smarterclayton** [comment]
Comments addressed, green [merge]

**Pull request merged**